build/invoke.go

```diff
@@ -0,0 +1,229 @@
+package build
+
+import (
+	"context"
+	_ "crypto/sha256" // ensure digests can be computed
+	"encoding/json"
+	"fmt"
+	"io"
+	"sync"
+	"sync/atomic"
+	"syscall"
+
+	"github.com/moby/buildkit/client"
+	"github.com/moby/buildkit/exporter/containerimage/exptypes"
+	gateway "github.com/moby/buildkit/frontend/gateway/client"
+	"github.com/moby/buildkit/solver/pb"
+	specs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+// InvokeConfig is configuration for a process to run in the container.
+type InvokeConfig struct {
```

Comment:
I think we should remove this struct, and instead use `controllerapi.InvokeConfig`, it would be good to avoid having multiple duplicate struct. The main difference seems to be that we obviously can't attach `io` objects to it, we can pass those as separate args (maybe using `ioset.In` to stick them together).

Note on having the duplicate `User` and `NoUser` fields to represent an optional value (which if we were using a modern version of protobuf, we could use the `optional` field in https://github.com/protocolbuffers/protobuf/blob/main/docs/field_presence.md). I guess there's not reason we have to use gogo proto for buildx like we do for buildkit (we don't care about version compatibility), we could just switch to using upstream? https://pkg.go.dev/google.golang.org/protobuf. Definitely follow-up material :tada: cc @crazy-max since you've done protobuf investigations.

Reply:
Thank you for the comment. Removed `InvokeConfig`.